daemon/logs.go

```diff
@@ -115,31 +144,55 @@ func (daemon *Daemon) ContainerLogs(ctx context.Context, containerName string, c
 }
 
 func (daemon *Daemon) getLogger(container *container.Container) (logger.Logger, error) {
-	if container.LogDriver != nil && container.IsRunning() {
-		return container.LogDriver, nil
+	if l := daemon.loggers.Get(container.ID); l != nil {
+		return l, nil
 	}
 	return container.StartLogger(container.HostConfig.LogConfig)
 }
 
+const loggerCloseTimeout = 10 * time.Second
+
 // StartLogging initializes and starts the container logging stream.
-func (daemon *Daemon) StartLogging(container *container.Container) error {
-	if container.HostConfig.LogConfig.Type == "none" {
+func (daemon *Daemon) StartLogging(c *container.Container) error {
+	if c.HostConfig.LogConfig.Type == "none" {
 		return nil // do not start logging routines
 	}
 
-	l, err := container.StartLogger(container.HostConfig.LogConfig)
+	l, err := daemon.getLogger(c)
 	if err != nil {
 		return fmt.Errorf("Failed to initialize logging driver: %v", err)
 	}
+	daemon.loggers.Add(c.ID, l)
 
-	copier := logger.NewCopier(map[string]io.Reader{"stdout": container.StdoutPipe(), "stderr": container.StderrPipe()}, l)
-	container.LogCopier = copier
-	copier.Run()
-	container.LogDriver = l
+	copier := logger.NewCopier(map[string]io.Reader{"stdout": c.StdoutPipe(), "stderr": c.StderrPipe()}, l)
 
+	go func() {
+		watch, cancel := daemon.watchEvent(&container.StateEvent{
```

Comment:
Oh, I have not seen this.
Ultimately though, I want to get rid of `state.waitChan`.